pkg/lib/kubernetes/pkg/util/hash/hash.go

```diff
@@ -17,21 +17,36 @@ limitations under the License.
 package hash
 
 import (
-	"hash"
-
-	"github.com/davecgh/go-spew/spew"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"math/big"
 )
 
 // DeepHashObject writes specified object to hash using the spew library
 // which follows pointers and prints actual values of the nested objects
 // ensuring the hash does not change when a pointer changes.
-func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) {
-	hasher.Reset()
-	printer := spew.ConfigState{
-		Indent:         " ",
-		SortKeys:       true,
-		DisableMethods: true,
-		SpewKeys:       true,
+func DeepHashObject(obj interface{}) (string, error) {
+	// While the most accurate encoding we could do for Kubernetes objects (runtime.Object)
+	// would use the API machinery serializers, those operate over entire objects - and
+	// we often need to operate on snippets. Checking with the experts and the implementation,
+	// we can see that the serializers are a thin wrapper over json.Marshal for encoding:
+	// https://github.com/kubernetes/kubernetes/blob/8509ab82b96caa2365552efa08c8ba8baf11c5ec/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L216-L247
+	// Therefore, we can be confident that using json.Marshal() here will:
+	//  1. be stable & idempotent - the library sorts keys, etc.
+	//  2. be germane to our needs - only fields that serialize and are sent to the server
+	//     will be encoded
+
+	raw, err := json.Marshal(obj)
+	if err != nil {
+		return "", fmt.Errorf("couldn't encode object: %w", err)
 	}
-	printer.Fprintf(hasher, "%#v", objectToWrite)
+
+	// base62(sha224(bytes)) is a useful hash and encoding for adding the contents of this
+	// to a Kubernetes identifier or other field which has length and character set requirements
+	hash := sha256.Sum224(raw)
```

Comment:
We could cut out the extra `raw` byte buffer if we use `json.NewEncoder(hash)`, I think?